### DISCUSSION TITLE
Estimate Multirotor Kinematics

### DISCUSSION THREAD

**Discussion by Sabrin0** [microsoft/AirSim#4214] [Support Q&A]
Hi, is there a way to estimate multirotor kinematics (position, orientation, linear and angular velocity) without actually move it on unreal?
For instance, to call the python method `moveByVelocity` and estimate the the drone navigation (for a limited amount of time) while keeping the same drone position in the unreal environment

**Comment by jonyMarino** [comment]
@Sabrin0. I can think of a way you could achieve that: First, save the kinematics state obtained by [simGetGroundTruthKinematics](https://microsoft.github.io/AirSim/api_docs/html/#airsim.client.VehicleClient.simGetGroundTruthKinematics) then call to move* and return to the previous kinematics state with [simSetKinematics](https://microsoft.github.io/AirSim/api_docs/html/#airsim.client.VehicleClient.simSetKinematics). The downside of this way is that this "reset" does not apply to the autopilot, so you cannot return to that point (It would be very interesting to change our autopilot (SimpleFlight) to support that type of rewind)
I can think of other ways, but I would like to know what you are trying to achieve to see if they would be correct.

**Reply by Sabrin0** [reply]
Hi @jonyMarino , thanks for you reply. Actually I would like to be able to control the drone while the kinematics estimation runs in _parallel_.
I noticed that the function `void PawnSimApi::updateKinematics` updates kinematics from pawn's movement. I'm thinking about creating a new class that inherits from `msr::airlib::Kinematics`. In this way I think that is possible to obtain all the data desired (position, orientation, linear and angular velocity). Therefore I'll create BP function library, which handles the computation whenever it's triggered (that's because I need the data inside UE).  So my question is: who manages the drone dynamics given  the client input (for instance desired velocities from `moveby*`)? 'cause I need to _clone_ that part of code, so store the data without moving the drone in the unreal environment.

**Reply by jonyMarino** [reply]
That is done inside FastPhysicsEngine.hpp.
You can separate the kinematics update there from the one inside PhysicsBody which is later used by the rest of the simulator

**Reply by Sabrin0** [reply]
Nice, thank you very much. I'll update you!

**Reply by Sabrin0** [reply]
I've looked at the` FastPhysicsEngine.hpp`, sorry but do you have any suggestion about how to calculate the next kinematics depending on the desired velocities?

**Reply by jonyMarino** [reply]
Ok, you don't only have the kinematics update but also the rotors' power update made by SimpleFlight. If you need the interaction between both then you will need a separate controller to call SimpleFlightApi::update() on it and update the rotor signals to then update the physics.

**Reply by Sabrin0** [reply]
I found another way, spawn a dummy drone to easily access to kinematics data. Thank you for your time @jonyMarino

**Reply by Sabrin0** [reply]
Hi @jonyMarino, sorry just one quick question: you told me to use the method `simSetKinematics`, but it does not exist even if there is in the documentation. My current version of airsim is 1.6

**Reply by jonyMarino** [reply]
Yes. It is not on v1.6 you will need to use the latest AirSim

**Comment by jonyMarino** [comment]
You need to clone the repository. Then you can use blocks or a custom
environment.

**Reply by Sabrin0** [reply]
Yes I followed the instruction  [Upgrading Your Custom Unreal Project](https://microsoft.github.io/AirSim/unreal_upgrade/#upgrading-your-custom-unreal-project), but now I'm no more able to run the simulation, both by building in Visual Studio (first image error) or by launching the .uproject

![VS_error](https://user-images.githubusercontent.com/73488155/146533847-86ce7de7-a664-4eec-90a8-2caf1a91a7e7.PNG)
![image](https://user-images.githubusercontent.com/73488155/146534204-0131a707-78be-499d-8aa7-10b1d063eb65.png)
( the version is the same as before: UE4.26)
![image](https://user-images.githubusercontent.com/73488155/146534330-998890cc-5836-44b2-8f94-ebc782123c24.png)

**Reply by Sabrin0** [reply]
Hi @jonyMarino, In order to solve the firs error I had to set as startup project the game from the VS solution explorer. By the way, new errors  jumped out:

```
Build started...
1>------ Build started: Project: DroneRescueNav, Configuration: DebugGame_Editor x64 ------
1>Using 'git status' to determine working set for adaptive non-unity build (C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav).
1>Building DroneRescueNavEditor...
1>Using Visual Studio 2019 14.29.30038 toolchain (C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037) and Windows 10.0.19041.0 SDK (C:\Program Files (x86)\Windows Kits\10).
1>[Adaptive unity build] Excluded from AirSim unity file: AirBlueprintLib.cpp, RpcLibClientBase.cpp, RpcLibServerBase.cpp, MultirotorApiBase.cpp, MultirotorRpcLibClient.cpp, MultirotorRpcLibServer.cpp, CameraDirector.cpp, PawnSimApi.cpp, PIPCamera.cpp, SimHUD.cpp, SimModeBase.cpp, SimModeWorldBase.cpp, CarPawnSimApi.cpp, SimModeCar.cpp, ComputerVisionPawn.cpp, SimModeComputerVision.cpp, MultirotorPawnSimApi.cpp, WorldSimApi.cpp
1>Building 4 actions with 12 processes...
1>  [1/4] WorldSimApi.cpp
1>C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\WorldSimApi.cpp(970): error C2653: 'AirSimSettings': is not a class or namespace name
1>C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\WorldSimApi.cpp(970): error C3861: 'singleton': identifier not found
1>C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\WorldSimApi.cpp(969): error C2660: 'msr::airlib::EarthUtils::nedToGeodetic': function does not take 1 arguments
1>  C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\AirLib\include\common/EarthUtils.hpp(291): note: see declaration of 'msr::airlib::EarthUtils::nedToGeodetic'
1>C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\WorldSimApi.cpp(972): error C3861: 'Vector2r': identifier not found
1>C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\WorldSimApi.cpp(973): error C3861: 'Vector2r': identifier not found
1>C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\WorldSimApi.cpp(979): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
1>C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\WorldSimApi.cpp(979): error C2143: syntax error: missing ';' before '&'
1>C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\WorldSimApi.cpp(979): error C2065: 'orientation': undeclared identifier
1>C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\Plugins\AirSim\Source\WorldSimApi.cpp(981): error C2065: 'orientation': undeclared identifier
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.MakeFile.Targets(45,5): error MSB3073: The command ""C:\Program Files\Epic Games\UE_4.26\Engine\Build\BatchFiles\Build.bat" DroneRescueNavEditor Win64 DebugGame -Project="C:\Users\Luca\Desktop\University\Thesis\DroneRescueNav\DroneRescueNav.uproject" -WaitMutex -FromMsBuild" exited with code 6.
1>Done building project "DroneRescueNav.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

**Comment by Sabrin0** [comment]
I created a new issue since maybe someone faces this problem too
